app/src/main/java/com/kickstarter/viewmodels/CommentsViewHolderViewModel.kt

```diff
@@ -199,48 +223,72 @@ interface CommentsViewHolderViewModel {
                     this.commentCardStatus.onNext(CommentCardStatus.RE_TRYING_TO_POST)
                 }
 
-            val commentData = this.commentInput.map {
-                Pair(requireNotNull(it.comment?.body()), requireNotNull(it.project))
-            }
+            comment
+                .compose(takeWhen(this.onFlagButtonClicked))
+                .compose(bindToLifecycle())
+                .subscribe(this.flagComment)
+        }
 
-            Observable
-                .combineLatest(postNewComment, commentData) { _, commentData ->
-                    return@combineLatest executePostCommentMutation(commentData)
-                }
+        /**
+         * Handles the logic for posting comments (new ones, and the retry attempts)
+         * @param commentData will emmit only in case we need to post a new comment
+         */
+        private fun postComment(commentData: Observable<Pair<String, Project>>) {
+            commentData
+                .map { executePostCommentMutation(it) }
                 .switchMap {
                     it
-                }.subscribe {
+                }
+                .compose(bindToLifecycle())
+                .subscribe {
                     this.commentCardStatus.onNext(CommentCardStatus.COMMENT_FOR_LOGIN_BACKED_USERS)
                 }
 
             Observable
-                .combineLatest(onRetryViewClicked, commentData) { _, commentData ->
-                    return@combineLatest executePostCommentMutation(commentData)
+                .combineLatest(onRetryViewClicked, commentData) { _, newData ->
+                    return@combineLatest executePostCommentMutation(newData)
                 }
                 .switchMap {
                     it
                 }.doOnNext {
                     this.commentCardStatus.onNext(CommentCardStatus.POSTING_COMMENT_COMPLETED_SUCCESSFULLY)
```

Comment:
@hadia  I think this `CommentCardStatus.POSTING_COMMENT_COMPLETED_SUCCESSFULLY` can be removed same as the `doOnNext` but let me know what you think about it

Reply:
That's when the state to show the green check and Posted keyword